draft-ietf-tls-esni.md

```diff
@@ -446,33 +490,114 @@ to harmonize these to make sure that we maintain key separation.]]
+  connection with an "illegal_parameter" alert. The client then proceeds
+  with the connection as usual, verifying the certificate against the desired
+  name.
+
+- If the value is "esni_retry_request", the client proceeds with the handshake,
+  verifying the certificate against ESNIKeys.public_name as described in
+  {{verify-public-name}}. If the handshake completes successfully, the client
+  MUST abort the connection with an "esni_required" alert.
+
+  The client then processes the "retry_keys" field from the server's
+  "encrypted_server_name" extension. If one of the values used a version known
+  to the client, the client SHOULD retry the handshake with a new transport
+  connection, using that value to encrypt the SNI. If no value is applicable,
+  the client SHOULD retry with ESNI disabled.
+
+  The client MUST NOT use the server-provided retry keys until the handshake
+  completes successfully. On success, it MUST NOT overwrite the DNS-provided
+  keys with the retry keys. It MUST use the retry keys at most once and
+  continue offering DNS-provided keys for subsequent connections. This avoids
+  introducing a tracking vector, should a malicious server present
+  client-specific retry keys.
+
+- If the field contains any other value, the client MUST abort the connection
+  with an "illegal_parameter" alert.
+
+If the server negotiates an earlier version of TLS, or if it does not
+provide an "encrypted_server_name" extension in EncryptedExtensions, the
+client proceeds with the handshake, verifying the certificate against
+ESNIKeys.public_name as described in {{verify-public-name}}. The client MUST
+NOT enable the False Start optimization {{RFC7918}}. If the handshake completes
+successfully, the client MUST abort the connection with an "esni_required" alert.
+It then SHOULD retry the handshake with a new transport connection and ESNI
+disabled.
+
+[[TODO: Key replacement is significantly less scary than saying that ESNI-naive
+  servers bounce ESNI off. Is it worth defining a strict mode toggle in the ESNI
+  keys, for a deployment to indicate it is ready for that? ]]
+
+The client MUST NOT treat this as a signal to disable encrypted SNI until the
+handshake completes successfully.
+
+Clients SHOULD implement a limit on retries caused by "esni_retry_request" or
+servers which do not acknowledge the "encrypted_server_name" extension. If the
+client does not retry in either scenario, it MUST report an error to the
+calling application.
+
+### Verifying against the public name {#verify-public-name}
+
+When the server cannot decrypt or does not process the "encrypted_server_name"
+extension, it continues with the handshake using the cleartext "server_name"
+extension instead (see {{server-behavior}}). Clients that offer ESNI then
+verify the certificate with the public name, as follows:
+
+- If the server resumed a session or negotiated parameters not based on
+  certificates, the client MUST abort the connection with an illegal_parameter
```

Comment:
"parameters not based on certificates" should probably be worded differently.  I assume the intent is that you can't negotiate a PSK-only mode and need to pick something where the server authenticates with a certificate?

Reply:
Yeah. Otherwise we don't have a public name to check. :-) How about "did not negotiate certificate-based authentication"?

Reply:
I think we need to stay positive, "negotiated a session that did not use a certificate for authentication" -- otherwise we get confused if we have any sort of PSK+cert or other joint authentication scheme.

Reply:
(Also missed this. I clearly don't know how to use GitHub...)

Done.